File: src/Functional/Erlang/facade.erl
-module(facade).
-export([main/0]).

authenticate(User) -> "auth(" ++ User ++ ")".
reserve(Sku) -> "reserve(" ++ Sku ++ ")".
charge(Cents) -> "charge(" ++ integer_to_list(Cents) ++ ")".

checkout(User, Sku, Cents) ->
    authenticate(User) ++ ">" ++ reserve(Sku) ++ ">" ++ charge(Cents).

main() ->
    io:format("checkout=~s~n", [checkout("alice", "SKU-42", 499)]).

File: src/Functional/Clojure/facade.clj
(defn authenticate [user] (str "auth(" user ")"))
(defn reserve [sku] (str "reserve(" sku ")"))
(defn charge [cents] (str "charge(" cents ")"))

(defn checkout-facade [user sku cents]
  (str (authenticate user) ">" (reserve sku) ">" (charge cents)))

(println (str "checkout=" (checkout-facade "alice" "SKU-42" 499)))
